### PR TITLE
Adding in the ability to deal with multi-core

### LINKED ISSUE #421: Question/Suggestion for Microsoft.DependencyInjection.SolrNet
So from what I can tell the this library does not expect multiple cores at this point, correct me if I am wrong.  However, I was able to get a hacked version to work on my local machine, I did it like this.

```csharp
public class SolrPersonService : SolrServer<PersonSolrModel>
{
    public SolrPersonService(SolrPersonServer personServer, IReadOnlyMappingManager mappingManager, IMappingValidator _schemaMappingValidator)
        : base(personServer, mappingManager, _schemaMappingValidator) { }
}

public class SolrPersonServer : SolrBasicServer<PersonSolrModel>
{
    public SolrPersonServer(IConfiguration configuration, SolrPersonExecuter queryExecuter, ISolrDocumentSerializer<PersonSolrModel> documentSerializer, ISolrSchemaParser schemaParser, ISolrHeaderResponseParser headerParser, ISolrQuerySerializer querySerializer, ISolrDIHStatusParser dihStatusParser, ISolrExtractResponseParser extractResponseParser)
        : base(new AutoSolrConnection(configuration.GetConnectionString("connection")), queryExecuter, documentSerializer, schemaParser, headerParser, querySerializer, dihStatusParser, extractResponseParser) { }
}

public class SolrPersonExecuter : SolrQueryExecuter<PersonSolrModel>
{
    public SolrPersonExecuter(ISolrAbstractResponseParser<PersonSolrModel> resultParser, IConfiguration configuration, ISolrQuerySerializer querySerializer, ISolrFacetQuerySerializer facetQuerySerializer, ISolrMoreLikeThisHandlerQueryResultsParser<PersonSolrModel> mlthResultParser)
        : base(resultParser, new AutoSolrConnection(configuration.GetConnectionString("connection")), querySerializer, facetQuerySerializer, mlthResultParser) { }
}
```
And this got me thinking what if the ISolrConnection could be changed to also use a generic type so that when a developer specifies a model they are querying on the connection could change based on that type, for example what could be added is a wrapper class that deals with the generic idea for the connection.

```csharp
public interface ISolrConnection<T> : ISolrConnection {
}

public class BasicGenericConnection<T> : ISolrConnection<T> {
    public BasicGenericConnection(ISolrConnection connection) {
        // Save connection information in this wrapper for use in all the other classes that use the connection information
    }
}
```
A small tweak could be added to the current AddSolrNet for the global solr connection

```csharp
var connection = new AutoSolrConnection(cores.Single().Url);
//Bind single type to a single url, prevent breaking existing functionality
services.AddSingleton<ISolrConnection>(connection);

// Add New Generic configuration class connection
services.AddTransient(typeof(ISolrConnection<>), typeof(BasicGenericConnection<>));
```
This would allow for another method to be added into the ServiceCollectionExtensions class that could allow the user to bind a model to a specific core for example this could be added:
```csharp
public static IServiceCollection AddSolrCore<TModel>(this IServiceCollection services, string url)
{
    var connection = new BasicGenericConnection<TModel>(new AutoSolrConnection(url));
    services.AddTransient(typeof(ISolrConnection<TModel>), (service) => connection );
    return services;
}
```
Just wanted to pose this just in case it could help out with the multi-core problem.

### PR DESCRIPTION
**Pull request by tspayne87**
This is an attempt to deal with multi-core problem
Fixes #421

### CODE REVIEW

**Comment by gjunge** [comment]
@tspayne87 thanks for the PR! I'm currently on holidays, I will review it at the end of the month.

File: Microsoft.DependencyInjection.SolrNet/ServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -86,5 +88,19 @@ private static IServiceCollection AddSolrNet(this IServiceCollection services, I
...
         }
+
+        /// <summary>
+        /// Method to deal with adding a second core into Microsoft's dependency injection system.
+        /// </summary>
+        /// <param name="services">The dependency injection service.</param>
+        /// <param name="url">The url for the second core.</param>
+        /// <typeparam name="TModel">The type of model that should be used for this core.</typeparam>
+        /// <returns>The dependency injection service.</returns>
+        public static IServiceCollection AddSolrCore<TModel>(this IServiceCollection services, string url)
```
Comment (gjunge) [inline Microsoft.DependencyInjection.SolrNet/ServiceCollectionExtensions.cs]: I'm a little bit on the fence regarding the naming. 
It's a little confusing that:
1) I first have to call `AddSolrNet` (and already pass in a URL, but no type)
2) Then for the next core call `AddSolrCore<> ()` and pass in another URL with a type.

Perhaps we could go the same way as DbContext:
``` csharp
services.AddDbContext<SomeContext>(options => someoptions) ;
services.AddDbContext<OtherContext>(options => someotheroptions);
```

So in our case:
Legacy:
`public static IServiceCollection AddSolrNet(this IServiceCollection services, string url)`
New:
` public static IServiceCollection AddSolrNet<TModel>(this IServiceCollection services, string url)`

I would like to see that calling the new method, will also work without the need to call the legacy one. 
So we should take the 'general DI setup' out of the legacy method, and into a private helper, and make sure inside that helper method that it does its work only once.

So the goal is that in my `startup.cs` I would be able to write:

services.AddSolrNet<Foobar>("http://solr.com/foobar");
services.AddSolrNet<Testing>("http://solr.com/testing");

**Comment by tspayne87** [comment]
Added in suggested change now you would do something like this.

```csharp
services.AddSolrNet("http://solr.com/foobar");
services.AddSolrNet<TestingModel>("http://solr.com/testing");
```

However, calling the AddSolrNet without a generic class for a second time would result in nothing being changed or added.  Should I just remove the AddSolrNet option or just allow for that to happen and just put a note in for that.

**Comment by gjunge** [comment]
@tspayne87 could perhaps throw an `InvalidOperationException` in that case

**Comment by tspayne87** [comment]
That would be a good idea, you got a specific message in mind?

**Comment by gjunge** [comment]
Something like:

> Only one non-typed Solr Core can be registered. In order to use multiple cores, use the typed AddSolrNet<>() overload.

Although `non-typed` is not the best wording, thought before about `generic` but that is confusing. Open to suggestions though, and I think that whatever we type will already help to user understand what is wrong.

**Comment by tspayne87** [comment]
Added in the exception operations, also I have been thinking about the class ```SolrCore```.  Since we are breaking the cores up into separate overloaded methods I do no think that it is needed anymore since we are just using it to store the url and nothing else.  Just wanted to ask before removing the code.

**Comment by gjunge** [comment]
Yeah, thats a good idea.

**Comment by tspayne87** [comment]
I went ahead and removed the SolrCore class and added in some more documentation to the methods.  Was there anything else you can think of that would be good to add or remove?

**Comment by gjunge** [comment]
If it wouldn't be too much of a hassle , could you add an example of the new usage to here:
https://github.com/SolrNet/SolrNet/blob/master/Documentation/Initialization.md#microsoftextensionsdependencyinjection-aspnet-core 
(Documentation/Initialization.md)

And here:
https://github.com/SolrNet/SolrNet/blob/master/Documentation/Multi-core-instance.md

That would really advertise your solution to the users. 

Thanks a million!

**Comment by tspayne87** [comment]
Added in some documentation to the two files you suggested, sorry it took so long been bogged down pretty hard with work.

**Comment by gjunge** [comment]
@tspayne87  Thanks a million!

**Pull request merged**